File: android/src/main/kotlin/xyz/adscope/adscope_sdk/manager/AMPSEventManager.kt
package xyz.adscope.adscope_sdk.manager

import android.app.Activity
import xyz.adscope.adscope_sdk.data.InitMethodNames
import xyz.adscope.adscope_sdk.data.SplashMethodNames
import xyz.adscope.adscope_sdk.data.InterstitialMethodNames
import xyz.adscope.adscope_sdk.data.NativeMethodNames
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import xyz.adscope.adscope_sdk.data.BannerMethodNames
import xyz.adscope.adscope_sdk.data.DrawMethodNames
import xyz.adscope.adscope_sdk.data.RewardedVideoMethodNames
import xyz.adscope.adscope_sdk.utils.FlutterPluginUtil
import java.lang.ref.WeakReference

class AMPSEventManager private constructor() : MethodCallHandler {

    private var channel: MethodChannel? = null
    private var mContext: WeakReference<Activity>? = null // 在 Android 中通常使用 Context

    companion object {
        private var sInstance: AMPSEventManager? = null
        fun getInstance(): AMPSEventManager {
            return sInstance ?: synchronized(this) {
                sInstance ?: AMPSEventManager().also { sInstance = it }
            }
        }
    }

    /**
     * 初始化 MethodChannel 并设置回调处理器
     * @param binaryMessenger Flutter引擎的BinaryMessenger
     */
    fun init(binaryMessenger: BinaryMessenger) {
        if (channel == null) {
            channel = MethodChannel(binaryMessenger, "adscope_sdk") // "amps_sdk" 是通道名称
            channel?.setMethodCallHandler(this) // 将当前类设置为回调处理器
        }
    }

    /**
     * 处理来自 Flutter 的方法调用
     */
    override fun onMethodCall(call: MethodCall, result: Result) {
        when {
            InitMethodNames.contains(call.method) -> {
                AMPSSDKInitManager.getInstance().handleMethodCall(call, result)
            }
            SplashMethodNames.contains(call.method) -> {
                AMPSSplashManager.getInstance().handleMethodCall(call, result)
            }
            InterstitialMethodNames.contains(call.method) -> {
                AMPSInterstitialManager.getInstance().handleMethodCall(call, result)
            }

            NativeMethodNames.contains(call.method) -> {
                AMPSNativeManager.getInstance().handleMethodCall(call, result)
            }
            RewardedVideoMethodNames.contains(call.method) -> {
                AMPSRewardedVideoManager.getInstance().handleMethodCall(call, result)
            }
            BannerMethodNames.contains(call.method) -> {
                AMPSBannerManager.getInstance().handleMethodCall(call, result)
            }

            DrawMethodNames.contains(call.method) -> {
                AMPSDrawManager.getInstance().handleMethodCall(call, result)
            }

            else -> {
                result.notImplemented() // 如果方法名未被识别
            }
        }
    }

    /**
     * 从原生端向 Flutter 发送消息
     * @param method 方法名
     * @param args 参数，可以是 null 或任何 Flutter 支持的类型
     */
    fun sendMessageToFlutter(method: String, args: Any?) {
        FlutterPluginUtil.runOnUiThread {
            channel?.invokeMethod(method, args)
        }
    }

    /**
     * 释放资源，清除 MethodChannel 的回调处理器和 Context
     */
    fun release() {
        channel?.setMethodCallHandler(null)
        channel = null // 可选，如果不再需要这个channel实例
        mContext = null
    }
}

File: ios/Classes/common/AMPSEventManager.swift
//
//  AMPSEventManager.swift
//  amps_sdk
//
//  Created by 王飞 on 2025/10/21.
//

import Foundation
import Flutter


class AMPSEventManager : NSObject{
   
    static let shared = AMPSEventManager()
    private override init(){super.init()}
    
    var channel: FlutterMethodChannel?
    var registrar: FlutterPluginRegistrar?
    func regist(_ registrar: FlutterPluginRegistrar?) {
        guard let registrar = registrar else {
            return
        }
        self.registrar = registrar
        channel = FlutterMethodChannel(name: "adscope_sdk", binaryMessenger:  registrar.messenger())
        channel?.setMethodCallHandler { methodCall, result in
            switch methodCall.method {
            case let name where  initMethodNames.contains(name):
                AMPSSDKInitManager.shared.handleMethodCall(methodCall, result: result)
            case let name where splashMethodNames.contains(name):
                AMPSSplashManager.shared.handleMethodCall(methodCall, result:result)
            case let name where interstitialMethodNames.contains(name):
                AMPSInterstitialManager.shared.handleMethodCall(methodCall, result: result)
            case let name where  nativeMethodNames.contains(name):
                AMPSNativeManager.shared.handleMethodCall(methodCall, result: result)
            case let name where  rewardVideoMethodNames.contains(name):
                AMPSRewardVideoManager.shared.handleMethodCall(methodCall, result: result)
            case let name where  bannerMethodNames.contains(name):
                AMPSBannerManager.shared.handleMethodCall(methodCall, result: result)
            case let name where  drawMethodNames.contains(name):
                AMPSDrawManager.shared.handleMethodCall(methodCall, result: result)
            default:
                result(FlutterMethodNotImplemented)
            }
        }
        
    }
    func sendToFlutter(_ method:String,arg:Any? = nil){
        DispatchQueue.main.async { [weak self] in
            self?.channel?.invokeMethod(method, arguments: arg)
        }
    }
    
    func getImage(_ name:String) -> UIImage?  {
        let imageName = name.components(separatedBy: "/").last
        let source = imageName?.components(separatedBy: ".").first
        let type = imageName?.components(separatedBy: ".").last ?? "png"
        
        let tem = "flutter_assets/" + name
        var arr1 = tem.components(separatedBy: "/")
        arr1.removeLast()
        let dir = arr1.joined(separator: "/")
        
        guard let frameworkPath = Bundle.main.path(forResource: "App", ofType: "framework", inDirectory: "Frameworks"),
              let bundle = Bundle(path: frameworkPath) else {
            return nil
        }
        
        // 根据屏幕 scale 优先选择对应分辨率的变体目录，不存在时回退到低分辨率
        let scale = UIScreen.main.scale
        let variants: [String]
        if scale >= 3 {
            variants = ["3.0x", "2.0x", ""]
        } else if scale >= 2 {
            variants = ["2.0x", ""]
        } else {
            variants = [""]
        }
        for variant in variants {
            let variantDir = variant.isEmpty ? dir : dir + "/" + variant
            if let imagePath = bundle.path(forResource: source, ofType: type, inDirectory: variantDir),
               let image = UIImage(contentsOfFile: imagePath) {
                return image
            }
        }
        return nil
    }
}
